docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,453 @@
+
+```python
+@functools.partial(jax.jit, static_argnums=(4, 5))
+def train_step(opt_state, variables, inputs, labels, apply_fn, tx_update_fn):
+
+  def loss_fn(params):
+    logits, new_model_state = apply_fn(
+        variables.copy(dict(params=params)), inputs, mutable=['batch_stats'])
+    loss = xent_loss(logits, labels)
+    return loss, new_model_state
+
+  variables, params = variables.pop('params')
+  (loss, new_model_state), grads = jax.value_and_grad(loss_fn, has_aux=True)(
+      params)
+  updates, new_opt_state = tx_update_fn(grads, opt_state, params)
+  new_params = optax.apply_updates(params, updates)
+  new_variables = variables.copy(dict(params=new_params, **new_model_state))
+  return new_opt_state, new_variables, loss
+
+
+opt_state = tx.init(variables['params'])
+for batch in ds.as_numpy_iterator():
+  opt_state, variables, loss = train_step(
+      opt_state, variables, batch['image'], batch['label'], model.apply,
+      tx.update)
+  print(loss)
+```
+
+Remarks:
+
+- Since `tx.update()` only transforms the gradient, we still need to call
+  `optax.apply_updates()` to apply these transformed gradients to the
+  parameters.
+- Compared with the [previous API], we can now keep the entire `variables`
+  including the `params` as an input and output to the `train_step()`.
+- Splitting `params` from `variables` is still necessary inside the train step
+  because we only want to compute gradients with respect to `params` and not the
+  entire `variables`.
+- We can still log internal optimizer state, such as the learning rate, as long
+  as Optax transformations expose that information in their respective state.
+  For example, `optax.scale_by_schedule()` currently only exposes
+  `opt_state.count` but could easily be extend to also expose the `step_size`.
+  The same is true for internal optimizer states that change over time.
+
+## Multi Optimizer
+[Multi Optimizer]: #multi-optimizer
+
+The [previous API] defined `flax.optim.MultiOptimizer` for processing different
+parts of the parameter tree with different optimizers:
+
+```python
+biases_traversal = flax.optim.ModelParamTraversal(
+    lambda path, _: path.endswith('/bias'))
+not_biases_traversal = flax.optim.ModelParamTraversal(
+    lambda path, _: not path.endswith('/bias'))
+
+optimizer_def = flax.optim.MultiOptimizer(
+    (biases_traversal, flax.optim.GradientDescent(learning_rate=0.1)),
+    (not_biases_traversal, flax.optim.GradientDescent(learning_rate=0.05)),
+)
+```
+
+Note how we first define a traversal that selects parameters based on their
+path (which is the concatenation of module scopes and variable name), and then
+create a `MultiOptimizer` that binds a different optimizer for each of these
+separate traversals.
+
+Optax has recently implemented `optax.masked()` that can be used for specifying
+gradient transformations that only applied to a subset of the gradients:
+
+```python
+def dict_map_with_path(fn, d, path=''):  # a crude "ModelParamTraversal"
```

Comment:
nit: this breaks our current path repr as a tuple of names

A simplified implementation using flatten_dict could be:
```
def dict_map_with_path(fn, data):
  flat = traverse_util.flatten_dict(data)
  return traverse_util.unflatten_dict({k: fn(k, v) for k, v in flat.items()})
```

Reply:
Updated to use Optax's new `mask=lambda ...` function & to adhere to Flax's "a path is a tuple" convention.